trunk/source/doc/gexec.1
.\"
.\" Source Version: 3.0
.\" Software Release #: LLNL-CODE-422942
.\"
.\" include cpyright.h
.\"

.TH GEXEC 1 "02 September 2012"
.SH NAME
gexec \- generalization of process pipelines
.SH SYNOPSIS
.B gexec [-cev] [-csv] [-d #] [-h] [-nv] [-p] [-s]
         [-sev] [-ssv] [-st #] [-t] <command>*

.SH DESCRIPTION

gexec runs groups of processes with more arbitrary I/O connectivity.
The standard UNIX shells have the ability of running pipelines of
process in which the output of one serves as the input to the next.
There are a few problems and limitations with the simple pipeline
model as implemented.

1) The exit status of all processes but the last is lost

2) The ability to redirect stdout to one process and stderr
   to another is absent

3) There is no fan out or fan in in a pipeline

4) Shell functions cannot interact with processes

5) Sending other information such as environment variables,
   resource limits, resource usage, or even exit status is
   not done

gexec is implemented as a C function which could be incorporated
into a code as an alternative to the system function.  It is available
in the PACT SCHEME interpreter where processes and SCHEME functions
can communicate.

The grammar for commands is:

     <command>   := <pgrp> | <pgrp> <st-sep> <command>
     <st-sep>    := ';'  |  '&&'  |  '||'

     <pgrp>      := <job> | <job> <pgrp-io> <pgrp>
     <job>       := <program> <args> | <program> <args> <file-io>
     <program>   := 'executable program or script'
     <args>      := <arg> | <arg> <args>
     <arg>       := 'argument to an executable'

     <file-io>   := <shell-io> | <pgr-io>
     <shell-io>  := <shell-op> | <shell-op> <file-name>
     <file-name> := 'name of file'
     <shell-op>  := '<' | '>' | '>&' | '&>' | 
                    '>>' | '>>&' |
                    '>!' | '>&!' |
                    '1>' | '2>' | '2>&1'

     <pgrp-io>   := <pipe-io> | <pgr-io>
     <pipe-io>   := '|'  |  '|&'  | '&|'

     <pgr-io>    := <pgr-spec> | <pgr-spec> <pgr-io>
     <pgr-spec>  := '@' |
                    '@' <io-kind> |
                    '@' <io-kind> <io-id> |
                    '@' <io-kind> <io-id> <io-kind>
     <io-kind>   := 'i' | 'o' | 'e' | 'b' |
                    'r' | 'x' | 'l' | 'v'
     <io-id>     := +<n> | -<n> | <n>
     <n>         := unsigned integer value
     
Illegal cases are:

     @i<n>i       input to input
     @o<n>o, @e<n>e, @o<n>e, or @e<n>o    output to output
     @o<n> @e<n>  where <n> is the same   use @b<n> to express that

Gexec works with the PACT utility perdb - the persistent database.
Perdb is used by processes in place of the environment to provide
a persistent database without regard to parent/child process relationships.
In this way arbitrary process can share a database (which is really
what the environment is) in a more symmetric and friendly way.
In particular, gexec sets a variable 'gstatus' in the persisent
database which is a text string (as are environment variables) containing
the exit status of each process in the process group run by gexec.
See the perdb man pages for more details of its operation.


.SH OPTIONS

.TP
.B \-cev
Print setting gstatus as an environment variable in CSH
form - setenv gstatus "..."
.TP
.B \-csv
Print setting gstatus as a shell variable in CSH form.
form - set gstatus = ( ... )
.TP
.B \-d
Set the debugging level to see information on I/O connections.
.TP
.B \-h
This help message.
.TP
.B \-nv
Do not print expression setting gstatus.
.TP
.B \-p
Use pipes to communcate between processes.
.TP
.B \-s
Use sockets to communcate between processes.
.TP
.B \-sev
Print setting gstatus as an environment variable in SH
form - gstatus="...";export gstatus
.TP
.B \-ssv
Print setting gstatus as a shell variable in SH
form - gstatus="..."
.TP
.B \-st
Set the exit status bit mask.  The exit status of gexec
will be the sum of all non-zero exit status group members
covered by the mask.  For example, -st 3 would look for
non-zero exit statuses from the 1rst and 2nd processes in
the group.  The default is to use all members of the group.
.TP
.B \-t
Use PTYs to communcate between processes (not complete).

.SH EXAMPLES

The following examples are take from the gexec test suite.
The program osrc is a simple program that prints messages
to both its stdout and stderr.  It is used for reproducibility.
If you wish to run one of these examples doing an ls on some
files that exist and some that don't will work as well.

.B Processes

Here are several examples of communication between processes.
Notice that some standard shell cases are supported and their
special syntax must be quoted to avoid it being caught by
the shell.

# simple asynchronous processes

 gexec pwd @ date @ ls

# simple stdout variants

 gexec /bin/echo "foo" @o cat -n
 gexec /bin/echo "foo" @o+1 cat -n
 gexec /bin/echo "foo" @o2 cat -n
 gexec /bin/echo "foo" '|' cat -n

# simple stdout and stderr variants

 gexec osrc -t j4 @o+1 @e+2 cat -n @ cat -E
 gexec osrc -t j5 @o2 @e3 cat -n @ cat -E

# simple bonded stdout/stderr variants

 gexec osrc -t j6 @b cat -n
 gexec osrc -t j7 @b+1 cat -n
 gexec osrc -t j8 @b2 cat -n
 gexec osrc -t s2 '|' cat -n
 gexec osrc -t s3 '|&' cat -n

.B Functions

These are examples of redirecting process I/O to and from
functions or procedures.  The syntax: 'pw:test foo' means
that test is a procedure that will called with the argument
foo and written to; 'pr:test foo' means test is a procedure
that will be called with the argument foo and read from;
and 'pa:test foo' means that test is a procedure which will
be called with the argument foo and it will be appended to
(this may not be a useful distinction but it is made for
consistentcy with file I/O).

These examples of functions are only vaguely useful.  The
gexec program has one built in function for testing the
interaction between processes and functions.  The PACT
SCHEME interpreter test suite has better examples showing
how to have functions work with processes.

The builtin gexec function is called 'test' and it prints
some messages to which the argument is prepended.

 gexec /bin/echo "foo" @o pw:test bar

# simple stdin variants

 gexec cat @i pr:test bar

# simple append variants

 gexec /bin/echo "foo" @o2 pa:test p3out

# simple bonded variants

 gexec /bin/echo "foo" @b2 pw:test p4bond

# non-simple stdout variants

 gexec /bin/echo "foo" @o2 @e3 pw:test p5out @ pw:test p5err
 gexec osrc -t p6 @o2 @e3 pw:test p6out @ pw:test p6err

.B Files

These are examples of redirecting process I/O to and from
files.  The syntax: 'fw:file.dat' means that file.dat is
a file and it will be created and written to; 'fr:file.dat'
means that file.dat is a file whose content will be read
and sent on; and 'fa:file.dat' means that file.dat is
a file and its contents will be appended to.

# simple stdout variants

 gexec /bin/echo "foo" @o fw:file.dat

# simple stdin variants

 gexec cat @i fr:file.dat

# simple append variants

 gexec /bin/echo "foo" @o2 fa:file.dat

# simple bonded variants

 gexec /bin/echo "foo" @b2 fw:f4bond

# non-simple stdout variants

 gexec /bin/echo "foo" @o2 @e3 fw:f5out @ fw:f5err
 gexec osrc -t f6 @o2 @e3 fw:f6out @ fw:f6err

.B Variables

These are examples of redirecting process I/O to and from
environment variables.  The syntax: 'vw:foo' means
that foo is a variable that will be written to; 'vr:foo'
means test is a variable that will be read from;
and 'va:foo' means that test is a variable which will
be appended to.

# simple stdout variants

 gexec /bin/echo "foo" @o vw:ev1

# simple stdin variants

 gexec cat @i vr:ev1

# simple append variants

 gexec /bin/echo "foo" @o2 va:ev1

# simple bonded variants

 gexec /bin/echo "foo" @b2 vw:v4bond

# non-simple stdout variants

 gexec /bin/echo "foo" @o2 @e3 vw:v5out @ vw:v5err
 gexec osrc -t v6 @o2 @e3 vw:v6out @ vw:v6err

.B Special kinds

The special I/O kinds are @x for exit status, @r for
resource usage, @v for environment variable(s), and @l
for resource limits.  For @x and @r the results are written
to specified process after the job is run.  For @v and @l
the results are read by the specified process and set before
it is exec'd.

# send the exit status of echo to cat
 gexec /bin/echo "foo" @x cat -n

# send the resource usage of echo to cat
 gexec /bin/echo "foo" @r cat -n

Note that the resource usages are at least the POSIX
mandated ones.

# run printenv with the addition of the variable foo
 gexec /bin/echo "foo=bar" @v printenv foo

# run csh with the core dump size limit
 gexec /bin/echo "core=13000" @l /bin/csh -cf limit @o grep coredumpsize

Note the options for setting resource limits are (depending
on the OS):

   core   - maximum core file size (RLIMIT_CORE)
   as     - address space - virtual memory (RLIMIT_AS)
   cpu    - CPU seconds limit (RLIMIT_CPU)
   fsize  - file size limit (RLIMIT_FSIZE)
   nofile - file number limit (RLIMIT_NOFILE)
   nproc  - thread number limit (RLIMIT_RSS)
   stack  - stack size limit (RLIMIT_STACK)


.SH See also
.TP 
.B perdb.

.SH BUGS
.TP 
.B None known.



source/psh/gexec.c
/*
 * GEXEC.C - process group management
 *
 * #include "cpyright.h"
 *
 */

#include "common.h"
#include "libpsh.c"
#include "libdb.c"
#include "libpgrp.c"

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* DO_TEST - do the I/O work for the test function */

static int do_test(FILE **fio, char *tag)
   {int i, rv, ev;
    char t[BFLRG];
    char *p;
    static int count = 0;

    rv = 0;

    for (i = 0; rv == 0; i++)
        {if (feof(fio[0]) == TRUE)
	    rv = 1;
	 else
	    {p  = fgets(t, BFLRG, fio[0]);
             ev = errno;
	     if (p != NULL)
	        {count = 0;
		 fprintf(fio[1], "%s> out: %s", tag, p);}
	     else if (feof(fio[0]) == TRUE)
	        rv = 1;
	     else if (ev == EBADF)
	        rv = 2;
	     else
	        {fprintf(fio[1], "%s> gen: %d\n", tag, ++count);
		 rv = (count > 6);};

	     nsleep(10);};};

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* STR_TEST - function to test */

static int str_test(char *db, io_mode md, FILE **fio,
		    char *name, int c, char **v)
   {int rv;
    char *tag;

    rv  = 0;
    tag = v[0];

    if (md == IO_MODE_RO)
       block_fd(fileno(stdin), FALSE);

    rv = do_test(fio, tag);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* GEXEC_CONV - run conversation between processes
 *            - to test both functions and processes for GEXEC
 */

int gexec_conv(char *db, io_mode md, FILE **fio,
	       char *name, int c, char **v)
   {int rv, nx;
    char s[BFLRG];
    char *p, *side;

    nx = 5;
    rv = 0;
    side = v[0];

    if ((side != NULL) && (md != IO_MODE_NONE))
       {if (strcmp(side, "a") == 0)
	   {static int ia = 0;

	    while (rv == 0)
	       {if (ia <= nx)
		   {fprintf(stderr, "A sent: %d\n", ia);
		    fprintf(fio[1], "%d\n", ia);
		    ia++;};
		p = fgets(s, BFLRG, fio[0]);
		if (p == NULL)
		   break;
		else
		   {fprintf(stderr, "A recv: %s", s);
		    if (strcmp(s, "end\n") == 0)
		       {fprintf(stderr, "Conv side a concluded\n");
			rv = 1;};};};}

	else if (strcmp(side, "b") == 0)
	   {static int ib = 0;

	    for (; (ib < 2*nx) && (rv == 0); )
	        {p = fgets(s, BFLRG, fio[0]);
		 if (p == NULL)
		    break;
		 else
		    {fprintf(stderr, "B recv: %s", s);
		     if (ib >= nx)
		        {fprintf(stderr, "Conv side b concluded\n");
			 fprintf(fio[1], "end\n");
			 rv = 1;}
		     else
		        {fprintf(stderr, "B sent: %c\n", 'a' + ib);
		         fprintf(fio[1], "%c\n", 'a' + ib);
		         ib++;};};};};};

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* MAPS - map function name to procedure for strong function execution */

static PFPCAL maps(char *s)
   {PFPCAL f;

    f = NULL;

    if (strcmp(s, "var") == 0)
       f = gexec_var;
    else if (strcmp(s, "file") == 0)
       f = gexec_file;
    else if (strcmp(s, "test") == 0)
       f = str_test;
    else if (strcmp(s, "conv") == 0)
       f = gexec_conv;

    return(f);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* HELP - print help message */

static int help(void)
   {int rv;

    printf("\n");
    printf("Usage: gexec [-cev] [-csv] [-d #] [-h] [-nv] [-p] [-s]\n");
    printf("             [-sev] [-ssv] [-st #] [-t] <cmd>\n");
    printf("   cev    output status as environment variable in CSH form\n");
    printf("   csv    output status as shell variable in CSH form\n");
    printf("   d      debug level\n");
    printf("   h      this help message\n");
    printf("   nv     do not print output status\n");
    printf("   p      use pipes between processes (default)\n");
    printf("   s      use sockets between processes\n");
    printf("   sev    output status as environment variable in SH form\n");
    printf("   ssv    output status as shell variable in SH form\n");
    printf("   st     set exit status bit mask\n");
    printf("   t      use PTY's between processes (not available yet)\n");
    printf("   <cmd>  see gexec man page for details\n");
    printf("\n");

    rv = 1;

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* MAIN - start here */

int main(int c, char **v, char **env)
   {int i, rv;
    char *db, *s;
    process_group_state *ps;

    rv = 0;
    ps = get_process_group_state();
    if (ps != NULL)
       {db = getenv("PERDB_PATH");
	s  = getenv("GEXEC_DEBUG_LEVEL");

	ps->dbg_level   = (s == NULL) ? 0 : atol(s);
	ps->status_mask = -1;

	for (i = 1; i < c; i++)
	    {if (strcmp(v[i], "-cev") == 0)
	        ps->ofmt = GEX_CSH_EV;

	     else if (strcmp(v[i], "-csv") == 0)
	        ps->ofmt = GEX_CSH_SV;

	     else if (strcmp(v[i], "-d") == 0)
	        {ps->dbg_level = atol(v[++i]);
		 csetenv("GEXEC_DEBUG_LEVEL", "%d", ps->dbg_level);}

	     else if (strcmp(v[i], "-h") == 0)
	        {rv = help();
		 break;}

	     else if (strcmp(v[i], "-nv") == 0)
	        ps->ofmt = GEX_NONE;

	     else if (strcmp(v[i], "-p") == 0)
	        ps->medium = IO_DEV_PIPE;

	     else if (strcmp(v[i], "-s") == 0)
	        ps->medium = IO_DEV_SOCKET;

	     else if (strcmp(v[i], "-sev") == 0)
	        ps->ofmt = GEX_SH_EV;

	     else if (strcmp(v[i], "-ssv") == 0)
	        ps->ofmt = GEX_SH_SV;

	     else if (strcmp(v[i], "-st") == 0)
	        ps->status_mask = atoll(v[++i]);

	     else if (strcmp(v[i], "-t") == 0)
	        ps->medium = IO_DEV_PTY;

	     else
	        {rv = gexec(db, c-i, v+i, env, maps);
		 break;};};};

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/
